### PR TITLE
A14.2 : Test case for Wordle

### PR DESCRIPTION
**Pull request by Ajay07Metamation**

### CODE REVIEW

File: Program.cs [hunk 1]
```diff
@@ -1,7 +1,216 @@
...
-      static void Main (string[] args) {
-         Console.WriteLine ("Hello, World!");
+      private static void Main (string[] args) {
       }
    }
+   public class Wordle {
+      // Interface ---------------------------------
+      // Public interface routine to run the game
+
+      public Wordle (StreamWriter writer) {
```
Comment (varalakshmi-imm) [inline Program.cs:15]: This should be an additional constructor. You should not change the original class for which you are adding tests. I should be able to run wordle irrespective of tests.

File: Program.cs [hunk 2]
```diff
@@ -1,7 +1,216 @@
...
+      public bool GameOver => mSucceeded || mFailed;
+      bool mSucceeded;     // User succeeded in guessing the word
+      bool mFailed;        // User failed to guess the word
+      public bool IsTesting {
+         get => mIsTest;
+         set => mIsTest = value;
+      }
+      bool mIsTest = false;
+
+      public bool CheckTextFilesEqual (string f1, string f2) {
```
Comment (varalakshmi-imm) [inline Program.cs:45]: This method need not be part of wordle class. You could have added it in Test project itself.

File: Program.cs [hunk 3]
```diff
@@ -1,7 +1,216 @@
...
+         mWriter.Write (data);
+      }
+
+      // Print the final result
+      public void PrintResult () {
+         if (!IsTesting) Put (KBDX, RESULTY, DarkGray, new string ('_', 31));
+         mWriter.WriteLine ();
+         if (mSucceeded) {
+            if (IsTesting) {
+               PutToFile ($"You found the word in {mY} tries", Green);
```
Comment (varalakshmi-imm) [inline Program.cs:173]: You are calling PutToFile in Put itself. So why do you need to call here separately?
Moreover the string messages are repeated.

File: Program.cs [hunk 4]
```diff
@@ -1,7 +1,216 @@
...
+            Put (WindowWidth / 2 - 15, RESULTY + 2, Green, $"You found the word in {mY} tries");
+         } else {
+            if (IsTesting) {
+               PutToFile ($"Sorry - the word was {SecretWord}", Yellow);
+               return;
+            }
+            Put (WindowWidth / 2 - 13, RESULTY + 2, Yellow, $"Sorry - the word was {SecretWord}");
+         }
+         if (!IsTesting) Put (WindowWidth / 2 - 11, RESULTY + 4, White, "Press any key to quit");
+         //ReadKey ();
```
Comment (varalakshmi-imm) [inline Program.cs:185]: Why is this commented out?

File: Program.cs [hunk 5]
```diff
@@ -1,7 +1,216 @@
...
+            mGuess[y] = new string (array);
+         }
+      }
+
+      public void PutToFile (object data, ConsoleColor color) {
+         data = color switch {
+            Green => $"{{{data}}}",
+            Blue => $"[{data}]",
+            DarkGray => $"({data})",
+            Gray or White or Yellow => $" {data} ",
```
Comment (varalakshmi-imm) [inline Program.cs:161]:
```suggestion
            _ => $" {data} ",

```

File: Program.cs [hunk 6]
```diff
@@ -1,7 +1,216 @@
...
+            _ => "",
```
Comment (varalakshmi-imm) [inline Program.cs:162]: Not required

File: TestWordle/UnitTest1.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
+// Test cases for wordle game
+
+using Training;
+
+namespace TestWordle {
+   [TestClass]
+   public class TestWordle {
+
+      // Test the backspace key input
```
Comment (varalakshmi-imm) [inline TestWordle/UnitTest1.cs:9]:
```suggestion
      // Tests the backspace key input

```
Change such comments everywhere

File: TestWordle/UnitTest1.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+      /// <summary>Run the game and tests different key inputs</summary>
+      /// <param name="mInput">Input words to the game</param>
+      /// <param name="Key">Key to be entered after entering each word</param>
+      /// <param name="mRefFile">Reference file for different game states</param>
+      public void TestFiles (string[] mInput, ConsoleKey Key, string mRefFile) {
+         var mTestFile = "C:/etc/TestFile.txt";
+         var writer = new StreamWriter (mTestFile);
+         Wordle mGame = new Wordle (writer);
+         mGame.SelectWord ();
```
Comment (varalakshmi-imm) [inline TestWordle/UnitTest1.cs:37]: Why is this required?
You are setting a secret word anyways.

File: TestWordle/UnitTest1.cs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      // Test the game by giving the correct word
+      [TestMethod]
+      public void Testsucceeded () =>
+         TestFiles (new string[] { "AFTER", "ABOUT", "APPLE" }, ConsoleKey.Enter, "Refdata/ReferenceSucceeded.txt");
+
+      /// <summary>Run the game and tests different key inputs</summary>
+      /// <param name="mInput">Input words to the game</param>
+      /// <param name="Key">Key to be entered after entering each word</param>
+      /// <param name="mRefFile">Reference file for different game states</param>
+      public void TestFiles (string[] mInput, ConsoleKey Key, string mRefFile) {
```
Comment (varalakshmi-imm) [inline TestWordle/UnitTest1.cs:33]:
```suggestion
      public void TestFiles (string[] input, ConsoleKey key, string refFile) {

```
Parameters should not have m prefix and they should start in lowercase.

File: TestWordle/UnitTest1.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+         var mTestFile = "C:/etc/TestFile.txt";
```
Comment (varalakshmi-imm) [inline TestWordle/UnitTest1.cs:34]:
```suggestion
         var testFile = "C:/etc/TestFile.txt";

```
Local variable too should not have m prefix only member variables.

File: TestWordle/UnitTest1.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+         mGame.SecretWord = "APPLE";
+         mGame.IsTesting = true;
+         foreach (var word in mInput) {
+            foreach (var ch in word)
+               mGame.UpdateGameState ((ConsoleKey)ch);
+            mGame.UpdateGameState (Key);
+         }
+         mGame.DisplayBoard ();
+         if (mGame.GameOver) mGame.PrintResult ();
+         writer.Close ();
```
Comment (varalakshmi-imm) [inline TestWordle/UnitTest1.cs:47]: Always use Using, so that the dispose will be called as sometimes we forget to call close.

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]
Not able to run Wordle application separately. 
Adding tests should not break the application's workflow.

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]
Exception thrown if I run wordle and enter a valid word like "HELLO"
image

File: TestWordle/UnitTest1.cs [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+               mGame.UpdateGameState (key);
+            }
+            mGame.DisplayBoard ();
+            if (mGame.GameOver) mGame.PrintResult ();
+         }
+         Assert.IsTrue (CheckTextFilesEqual (refFile, testFile));
+
+         // Returns true if the files match else shows the difference in winMerge 
+         bool CheckTextFilesEqual (string f1, string f2) {
+            var file1 = File.ReadAllText (f1).Replace ("\r\n", "\n");
```
Comment (varalakshmi-imm) [inline TestWordle/UnitTest1.cs:52]:
```suggestion
            var file1 = File.ReadAllText (f1);

```
Isn't this enough?

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

File: TestWordle/UnitTest1.cs (hunk 12 above)
Comment (Ajay07Metamation) [reply TestWordle/UnitTest1.cs:52]: It works but sometimes it opens winMerge even though the files are same.winMerge tells that the carraige is different but both files are identical.

Comment (varalakshmi-imm) [reply TestWordle/UnitTest1.cs]: Always replacing \r\n with \n. So this doesn't make sense to me if the reference files are also created from the test initially.

**Review by varalakshmi-imm — APPROVED** [review]

**Pull request closed without merging**